rpc/cc/arduino/cli/commands/v1/common.proto

```diff
@@ -108,6 +108,11 @@ message Platform {
   // If the platform is also not indexed it may fail to work correctly in some
   // circumstances, and it may need to be re-installed.
   bool missing_metadata = 14;
+  // Newest available version of the platform containing all compatible
+  // dependencies.
+  string latest_compatible = 15;
+  // True if the platform contains an incompatible dependency.
+  bool incompatible = 16;
```

Comment:
Hi, why not `compatible`? Is there a reason to invert the logic and introduce a  `latest_compatible` property together with the `incompatible`?

---

Please review the `Platform` type. The new field (no matter if we name as `compatible` or `incompatible`) will provide sufficient info to clients to determine if a historical revision of a platform retrieved by `core search --all` is compatible. But storing all this info on the same level is very strange. Attaching the `website`, `email`, `boards`, etc. properties to the historical revision of a platform is reasonable. But `installed`, `latest`, `latest_compatible`, etc., are unrelated to a revision and describe the currently managed state by the CLI.

Reply:
>Hi, why not compatible

I've used the `incompatible` as we can benefit from the default zero value of protobuf. Since most platform versions are compatible, that field will be false, and gRPC doesn't need to send that information to the wire. This was my rationale. But TBH I'd pick a better name since the performance of the payload transfer is negligible here. If you have some suggestions about it please share. (Some ideas: `installable`, `compatible`)

> Please review the Platform type

Thank you for bringing this up! I agree with you. I've discussed with @cmaglie some ideas. 
Modifying the Platform type means that we have to change also: `PlatformListResponse`.
Below are the proposed API changes.

```diff
diff --git a/rpc/cc/arduino/cli/commands/v1/common.proto b/rpc/cc/arduino/cli/commands/v1/common.proto
index cf452120..9f3cafe2 100644
--- a/rpc/cc/arduino/cli/commands/v1/common.proto
+++ b/rpc/cc/arduino/cli/commands/v1/common.proto
@@ -74,10 +74,10 @@ message Programmer {
 message Platform {
   // Platform ID (e.g., `arduino:avr`).
   string id = 1;
-  // Version of the platform.
+  // Installed version of the platform
   string installed = 2;
-  // Newest available version of the platform.
-  string latest = 3;
+  // Current version of the platform
+  string version = 3;
   // Name used to identify the platform to humans (e.g., "Arduino AVR Boards").
   string name = 4;
   // Maintainer of the platform's package.
@@ -108,11 +108,8 @@ message Platform {
   // If the platform is also not indexed it may fail to work correctly in some
   // circumstances, and it may need to be re-installed.
   bool missing_metadata = 14;
-  // Newest available version of the platform containing all compatible
-  // dependencies.
-  string latest_compatible = 15;
   // True if the platform contains an incompatible dependency.
-  bool incompatible = 16;
+  bool incompatible = 15;
 }

 message InstalledPlatformReference {
diff --git a/rpc/cc/arduino/cli/commands/v1/core.proto b/rpc/cc/arduino/cli/commands/v1/core.proto
index 70ec4d46..46aa6d1a 100644
--- a/rpc/cc/arduino/cli/commands/v1/core.proto
+++ b/rpc/cc/arduino/cli/commands/v1/core.proto
@@ -138,6 +138,12 @@ message PlatformListRequest {
 }

 message PlatformListResponse {
-  // The installed platforms.
-  repeated Platform installed_platforms = 1;
+  // List of platforms.
+  repeated PlatformListItem platforms = 1;
+}
+
+message PlatformListItem {
+  Platform platform = 1;
+  string latest = 2;
+  string latest_compatible = 3;
 }

```

Reply:
Actually, I would remove also the `installed` field from `Platform` (and maybe replicate it in `PlatformListItem`).

Reply:
I did not forget about this. Thank you!

> platform retrieved by `core search --all`

Could you please make a diff for `core search`?

Reply:
>Could you please make a diff for core search?

I'm not sure what I should diff against.
The diff I posted a couple of messages above (https://github.com/arduino/arduino-cli/pull/2285#discussion_r1334086979) 
is the only thing that changes from the latest commit I pushed and the proposed changes. All the rest of the API remains the same.

Here how it would look like the protobuf if we accept the proposed changes:
```protobuf
message PlatformListResponse {
  // List of platforms.
  repeated PlatformListItem platforms = 1;
}

message PlatformListItem {
  Platform platform = 1;
  string latest = 2;
  string latest_compatible = 3;
}

message Platform {
  // Platform ID (e.g., `arduino:avr`).
  string id = 1;
  // Installed version of the platform
  string installed = 2;
  // Current version of the platform
  string version = 3;
  // Name used to identify the platform to humans (e.g., "Arduino AVR Boards").
  string name = 4;
  // Maintainer of the platform's package.
  string maintainer = 5;
  // A URL provided by the author of the platform's package, intended to point
  // to their website.
  string website = 6;
  // Email of the maintainer of the platform's package.
  string email = 7;
  // List of boards provided by the platform. If the platform is installed,
  // this is the boards listed in the platform's boards.txt. If the platform is
  // not installed, this is an arbitrary list of board names provided by the
  // platform author for display and may not match boards.txt.
  repeated Board boards = 8;
  // If true this Platform has been installed manually in the user' sketchbook
  // hardware folder
  bool manually_installed = 9;
  // If true this Platform has been deprecated
  bool deprecated = 10;
  // Type of the platform.
  repeated string type = 11;
  // A URL provided by the author of the platform's package, intended to point
  // to their online help service.
  HelpResources help = 12;
  // If true the platform is indexed
  bool indexed = 13;
  // This field is true when the platform is installed with the Arduino IDE 1.8.
  // If the platform is also not indexed it may fail to work correctly in some
  // circumstances, and it may need to be re-installed.
  bool missing_metadata = 14;
  // True if the platform contains an incompatible dependency.
  bool incompatible = 15;
}
 ```

Reply:
Without assuming anything go, viper, etc. specific, this is my model change [proposal](https://github.com/arduino/arduino-cli/compare/master...kittaakos:arduino-cli:platform-grpc-api):
 - it pulls up the installation/versions/index/metadata props to a summary model,
 - a summary instance can have 1..* platform revisions.

Let me know what you think.

_(Note: I changed the model in a text editor. Hence, it might not be valid, but I hope it helps to get the idea.)_

Reply:
> I've used the `incompatible` as we can benefit from the default zero value of protobuf. Since most platform versions are compatible, that field will be false, and gRPC doesn't need to send that information to the wire.

+1 for a better API and using the `compatible` property instead of the `incompatible` one with this insignificant optimization.